### DISCUSSION TITLE
archery 在线查询记录查询审计日志

### DISCUSSION THREAD

**Discussion by woshiyanghai** [hhyo/Archery#2059] [Ideas]
1.archery 记录是在查询未超时 ，没有错误的情况下 ，才记录日志。
存在问题如下：
1）如果这次查询sql正确但是因为查询执行计划不好，执行时间过长，就会查询超时，web界面返回错误。但是查询审计日志并未记录此次查询，但是这次查询又真正的在数据库中执行过。

2为什么要记录这次查询超时的sql
1）这个sql确实在数据库执行过。可能导致cpu过高或者io过高等相关报警如果不记录，那么在后期去定位查询sql的时候比较不友好

**Comment by woshiyanghai** [comment]
# 仅将成功的查询语句记录存入数据库
        if not query_result.error:
            result['data']['seconds_behind_master'] = seconds_behind_master
            if int(limit_num) == 0:
                limit_num = int(query_result.affected_rows)
            else:
                limit_num = min(int(limit_num), int(query_result.affected_rows))
            query_log = QueryLog(
                username=user.username,
                user_display=user.display,
                db_name=db_name,
                instance_name=instance.instance_name,
                sqllog=sql_content,
                effect_row=limit_num,
                cost_time=query_result.query_time,
                priv_check=priv_check,
                hit_rule=query_result.mask_rule_hit,
                masking=query_result.is_masked
            )
            # 防止查询超时
            if connection.connection and not connection.is_usable():
                close_old_connections()
            query_log.save()
    except Exception as e:
        logger.error(f'查询异常报错，查询语句：{sql_content}\n，错误信息：{traceback.format_exc()}')
        result['status'] = 1
        result['msg'] = f'查询异常报错，错误信息：{e}'
        return HttpResponse(json.dumps(result), content_type='application/json')

**Comment by hhyo** [comment]
可以做下调整，提交查询执行前就记录，欢迎pr

**Reply by woshiyanghai** [reply]
查询之前由于无法拿到返回行数，和sql耗时，所以如果我们在查询前就记录，也不是很好。
不如直接在 查询超时之后，记录下这个sql信息就行了

**Reply by woshiyanghai** [reply]
# 仅将成功的查询语句记录存入数据库
        if not query_result.error:
            result['data']['seconds_behind_master'] = seconds_behind_master
            if int(limit_num) == 0:
                limit_num = int(query_result.affected_rows)
            else:
                limit_num = min(int(limit_num), int(query_result.affected_rows))
            # 防止查询超时
            if connection.connection and not connection.is_usable():
                close_old_connections()
        else:
            limit_num = 0
        query_result= QueryLog(
            username=user.username,
            user_display=user.display,
            db_name=db_name,
            instance_name=instance.instance_name,
            sqllog=sql_content,
            effect_row=limit_num,
            cost_time=query_result.query_time,
            priv_check=priv_check,
            hit_rule=query_result.mask_rule_hit,
            masking=query_result.is_masked
        )
        query_log.save()
